### ISSUE TITLE
lightseq inference abnormal using ls_fs_transformer_export.py exported model

### ISSUE THREAD

**Issue by dearchill** [bytedance/lightseq#207]
Hi, I used the _python export/ls_fs_transformer_export.py_ to export lightseq trained NMT model to do inference, but I found the result is quiet abnormal. These are some details output in the ls_fs_transformer_export.py test part.

> ***generator config***
beam size: 4
extra decode length(max decode length - src input length): 50
length penalty: 0.6
diverse lambda: 0
sampling method: beam_search
topk: 1
topp: 0.75
Allocated 882MB GPU buffer for transformer
decoder buffer init start
decoder buffer init succeed
pb results: (array([[[  4, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164,   6]]], dtype=int32), array([[0.]], dtype=float32))
hdf5 results: (array([[[  4, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164, 164,
         164,   6]]], dtype=int32), array([[0.]], dtype=float32))

I also tested more examples, and it continued to generate some repeated logits, and when I decoded the array with my tgt_dict, it generated something like this: 

> thesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesamesame.

I used the 0.10.2 fairseq and 2.1.4 lightseq version, and the lightseq-generate result seems normal. I think maybe something wrong happened in the export procedure. Looking forward for your reply.

**Comment by dearchill** [comment]
Can someone help me?

**Comment by godweiyang** [comment]
> Can someone help me?

Sorry to reply to you so late. Can you join our Feishu（飞书） group? Link: https://applink.feishu.cn/client/chat/chatter/add_by_link?link_token=0c2u75d7-5df4-4153-b436-39158d952ad8

We can communicate in the group chat detailly. You need not to download Feishu, because we have online service.

**Comment by godweiyang** [comment]
> --share-all-embeddings

You can not use `--share-all-embeddings`, and you should use `--share-decoder-input-output-embed`.

**Comment by dearchill** [comment]
@godweiyang Hi, I retrained the model using pre_ln, then exported the model, and the result seems normal, but still has some issues. The results are as follows:

```initializing tokenizer...
input texts: 回去 自己 煎药 还是 我们 煎 好 给 你 ？
[ 3013   110 31345  1741   415    23 30091    66    91    15    31     2]
creating lightseq model...
Parsing protobuf: ./export/transformer.pb
Finish loading src_emb_wei from host to device
Finish loading trg_emb_wei from host to device
Finish loading enc_wei from host to device
Finish loading dec_wei from host to device
Finish loading all weight from host to device
***model config***
encoder layers: 6
decoder layers: 6
hidden size: 512
inner size: 2048
head number: 8
dim per head: 64
src vocab size: 50000
trg vocab size: 50000
is_post_ln: 0
no_scale_embedding: 0
use_gelu: 0
start_id: 2
end_id: 6
padding_id: 2
multilg_type: 0

***generator config***
beam size: 4
extra decode length(max decode length - src input length): 50
length penalty: 1
diverse lambda: 0
sampling method: beam_search
topk: 1
topp: 0.75
Allocated 450MB GPU buffer for transformer
decoder buffer init start
decoder buffer init succeed
=========lightseq=========
lightseq generating...
(array([[[  143,   154,     9,    62,   220,  3225,    30,    44,    41,
         28063,    19,    14,    18,    34,    44,    41, 28063,    19,
            14,    18,    34,    44,    41, 28063,    19,    14,    18,
            34,    44,    41, 28063,    19,    14,    18,    34,    44,
            41, 28063,    19,    14,    18,    34,    44,    41, 28063,
            19,    14,    18,    34,    44,    41, 28063,    19,    14,
            18,    34,    44,    41, 28063,    19,     6]]], dtype=int32), array([[0.]], dtype=float32))
translation texts: go back to your own medicine or will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it.
```

Looks like it continues to generate repeated blocks. And another case goes like this:

```initializing tokenizer...
input texts: 众所周知 ， 人们 在 正式 场合 说话 更 有 礼貌 。
[7093    5  274    8 1565 4742 1859   62   22 5907    6    2]
creating lightseq model...
Parsing protobuf: ./export/transformer.pb
Finish loading src_emb_wei from host to device
Finish loading trg_emb_wei from host to device
Finish loading enc_wei from host to device
Finish loading dec_wei from host to device
Finish loading all weight from host to device
***model config***
encoder layers: 6
decoder layers: 6
hidden size: 512
inner size: 2048
head number: 8
dim per head: 64
src vocab size: 50000
trg vocab size: 50000
is_post_ln: 0
no_scale_embedding: 0
use_gelu: 0
start_id: 2
end_id: 6
padding_id: 2
multilg_type: 0

***generator config***
beam size: 4
extra decode length(max decode length - src input length): 50
length penalty: 1
diverse lambda: 0
sampling method: beam_search
topk: 1
topp: 0.75
Allocated 450MB GPU buffer for transformer
decoder buffer init start
decoder buffer init succeed
=========lightseq=========
lightseq generating...
(array([[[   75,    26,   768,     9,  1566,    57, 24317,    11,  3639,
          3053,     6,     6,     6]]], dtype=int32), array([[-2.228455]], dtype=float32))
translation texts: people are known to speak more politely in formal situations...
```

And it continues to generate the "."
I think maybe it's caused by the tokenizer inconsistency.  I can see in the export config, the start_id equals to 2, end_id equals to 6, and padding_id equals to 2. Is this fixed or related to my tokenizer?

**Comment by cc2017111** [comment]
> @godweiyang Hi, I retrained the model using pre_ln, then exported the model, and the result seems normal, but still has some issues. The results are as follows:
> 
> ```
> input texts: 回去 自己 煎药 还是 我们 煎 好 给 你 ？
> [ 3013   110 31345  1741   415    23 30091    66    91    15    31     2]
> creating lightseq model...
> Parsing protobuf: ./export/transformer.pb
> Finish loading src_emb_wei from host to device
> Finish loading trg_emb_wei from host to device
> Finish loading enc_wei from host to device
> Finish loading dec_wei from host to device
> Finish loading all weight from host to device
> ***model config***
> encoder layers: 6
> decoder layers: 6
> hidden size: 512
> inner size: 2048
> head number: 8
> dim per head: 64
> src vocab size: 50000
> trg vocab size: 50000
> is_post_ln: 0
> no_scale_embedding: 0
> use_gelu: 0
> start_id: 2
> end_id: 6
> padding_id: 2
> multilg_type: 0
> 
> ***generator config***
> beam size: 4
> extra decode length(max decode length - src input length): 50
> length penalty: 1
> diverse lambda: 0
> sampling method: beam_search
> topk: 1
> topp: 0.75
> Allocated 450MB GPU buffer for transformer
> decoder buffer init start
> decoder buffer init succeed
> =========lightseq=========
> lightseq generating...
> (array([[[  143,   154,     9,    62,   220,  3225,    30,    44,    41,
>          28063,    19,    14,    18,    34,    44,    41, 28063,    19,
>             14,    18,    34,    44,    41, 28063,    19,    14,    18,
>             34,    44,    41, 28063,    19,    14,    18,    34,    44,
>             41, 28063,    19,    14,    18,    34,    44,    41, 28063,
>             19,    14,    18,    34,    44,    41, 28063,    19,    14,
>             18,    34,    44,    41, 28063,    19,     6]]], dtype=int32), array([[0.]], dtype=float32))
> translation texts: go back to your own medicine or will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it for you? will we fry it.
> ```
> 
> Looks like it continues to generate repeated blocks. And another case goes like this:
> 
> ```
> input texts: 众所周知 ， 人们 在 正式 场合 说话 更 有 礼貌 。
> [7093    5  274    8 1565 4742 1859   62   22 5907    6    2]
> creating lightseq model...
> Parsing protobuf: ./export/transformer.pb
> Finish loading src_emb_wei from host to device
> Finish loading trg_emb_wei from host to device
> Finish loading enc_wei from host to device
> Finish loading dec_wei from host to device
> Finish loading all weight from host to device
> ***model config***
> encoder layers: 6
> decoder layers: 6
> hidden size: 512
> inner size: 2048
> head number: 8
> dim per head: 64
> src vocab size: 50000
> trg vocab size: 50000
> is_post_ln: 0
> no_scale_embedding: 0
> use_gelu: 0
> start_id: 2
> end_id: 6
> padding_id: 2
> multilg_type: 0
> 
> ***generator config***
> beam size: 4
> extra decode length(max decode length - src input length): 50
> length penalty: 1
> diverse lambda: 0
> sampling method: beam_search
> topk: 1
> topp: 0.75
> Allocated 450MB GPU buffer for transformer
> decoder buffer init start
> decoder buffer init succeed
> =========lightseq=========
> lightseq generating...
> (array([[[   75,    26,   768,     9,  1566,    57, 24317,    11,  3639,
>           3053,     6,     6,     6]]], dtype=int32), array([[-2.228455]], dtype=float32))
> translation texts: people are known to speak more politely in formal situations...
> ```
> 
> And it continues to generate the "." I think maybe it's caused by the tokenizer inconsistency. I can see in the export config, the start_id equals to 2, end_id equals to 6, and padding_id equals to 2. Is this fixed or related to my tokenizer?

Hello, I have the same problem as you. Can I have a look at your FS training script? How to add pre layernormalization in FS training script?

**Comment by dearchill** [comment]
Add these args: --encoder-normalize-before --decoder-normalize-before

**Comment by dearchill** [comment]
start_id=2, end_id=6, padding_id=1 solved my problem. Closed it.

**Comment by dearchill** [comment]
start_id=2, end_id=2, padding_id=1 solved my problem. Closed it.
